lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -411,12 +411,18 @@ def is_satisfied_by(self, requirement, candidate):
         # NOTE: pre-releases is when there are several user requirements
         # NOTE: and one of them is a pre-release that also matches a
         # NOTE: transitive dependency of another requirement.
-        allow_pre_release = self._with_pre_releases or not (
-            requirement.ver == '*' or
-            requirement.ver.startswith('<') or
-            requirement.ver.startswith('>') or
-            requirement.ver.startswith('!=')
-        ) or self._is_user_requested(candidate)
+        # NOTE: Finally, if a concrete artifact is met, like a Git repository,
+        # NOTE: subdirs, a tarball URL, or a local dir or file etc,
+        # NOTE: pre-releases are allowed regardless of any other circumstances.
+        allow_pre_release = (
+            self._with_pre_releases or
+            candidate.is_concrete_artifact or not (
```

Comment:
I think this is okay, though it is much broader than `type: git`, and we don't have much test coverage for dependencies that are concrete artifacts as it pertains to pre-releases. At least we know the missing test coverage/potential issues of the change is limited to dependencies, since direct reqs were handled by the last `or`.

If this does have any unintended side effects, the only workaround would probably be `--no-deps`.

Reply:
I meant for this to apply beyond Git as while making this patch, I recognized that the problem would manifest in each of those cases otherwise. So I've gone for a generic solution. But you're right, we should have better coverage.

Reply:
I was questioning the additional change on top of that regarding dependencies (as the lay-user will see them). The generic fix is a bit redundant, but once the dead code is removed the simplification definitely seems like a net positive if it doesn't have side effects.

Reply:
Reverted. Marking this as resolved.

Reply:
I don't think it's dead. It just covers a different part of the corner case.

Reply:
Alright.. After more testing and tweaks, I mark this thread as resolved.